karen/models/UNet.py

```diff
@@ -0,0 +1,144 @@
+import torch
+import torch.nn as nn
+from ..base_model import BaseModel
+from ..register_model import RegisterModel
+from math import log2
+
+
+def convnorm(in_ch, out_ch, ks):
+    return nn.Sequential(
+        nn.Conv1d(in_ch, out_ch, ks, padding=ks // 2),
+        nn.BatchNorm1d(out_ch),
+        nn.LeakyReLU(0.1),
+    )
+
+
+@RegisterModel("UNet")
+class UNet(BaseModel):
+    def __init__(
+        self,
+        embeddings,
+        in_feat,
+        out_feat,
+        unet_depth,
+        channel_jump,
+        encoding_ks,
+        decoding_ks,
+    ):
+        super(UNet, self).__init__()
+        self.embeddings = embeddings
+        channels_in = self.embeddings.weight.shape[-1]
+
+        self.init_linear_converters(unet_depth, in_feat, out_feat)
+        self.init_channel_sizes(channel_jump, unet_depth, channels_in)
+
+        self.depth = unet_depth
+
+        self.encoder = nn.ModuleList()
+        self.decoder = nn.ModuleList()
+        for i in range(self.depth):
+            self.encoder.append(convnorm(self.echannelin[i], self.echannelout[i], encoding_ks))
+            self.decoder.append(convnorm(self.dchannelin[i], self.dchannelout[i], decoding_ks))
+
+        self.middle = convnorm(self.echannelout[-1], self.echannelout[-1], encoding_ks)
+
+        self.out = nn.Sequential(
+            nn.Conv1d(channels_in + (channels_in // channel_jump + 1) * channel_jump, 1, 1),
+            nn.Tanh(),
+        )
+
+        self.upsample = nn.Upsample(scale_factor=2, mode="linear", align_corners=False)
+        self.downsample = nn.MaxPool1d(2)
+
+    def init_linear_converters(self, unet_depth, in_feat, out_feat):
+        linear_size = max(int(2 ** unet_depth), int(2 ** int(log2(in_feat))))
+        self.conversion_lin_in = nn.Linear(in_feat, linear_size)
+        self.conversion_lin_out = nn.Linear(linear_size, out_feat)
+        print(
+            "Unet needs to convert the sentence length to max({}, {}) in order to work correctly\n".format(
+                int(2 ** unet_depth), int(2 ** int(log2(in_feat)))
+            )
+        )
```

Comment:
Why?

Reply:
The input size is 165, so the convolutions become quite complicated because it's not a power of 2. So we convert to the nearest power of 2. In addition, because of the depth of the model and the fact that each layer uses max pooling, if the size of the input is not big enough the output of max pooling removes a dimension from the data as it reduces the size too much. (Dividing by 2 eventually gives 0) so to avoid this we need to compute the max of the nearest power of 2, and 2^depth.